### PR TITLE
[Fix] Weight calculations for Wild -> affects teleports

### PR DESCRIPTION
**Pull request by ruseinov**
Fixes issues with high teleport costs due to high multiplier for `Wild` assets, which is always the case with teleports. 

This is blocking the current release.

### CODE REVIEW

File: parachains/runtimes/assets/statemine/src/weights/xcm/mod.rs [hunk 1]
```diff
@@ -31,13 +31,12 @@ trait WeighMultiAssets {
 	fn weigh_multi_assets(&self, weight: Weight) -> XCMWeight;
 }
 
-const MAX_ASSETS: u32 = 100;
+const MAX_ASSETS: u32 = 1;
```
[the remark this answers is not in the document]
Comment (ruseinov) [reply parachains/runtimes/assets/statemine/src/weights/xcm/mod.rs]: This does not concern collective runtimes.

**Review by muharem — APPROVED** [review]

**Comment by ruseinov** [comment]
@gilescope FYI I've ran the integration tests and teleports are now about 12 times cheaper than they were before those weights were introduced, instead of being 20 times more expensive.

File: parachains/common/src/xcm_config.rs [hunk 2]
```diff
@@ -119,3 +119,17 @@ impl<Location: Get<MultiLocation>> FilterAssetLocation for ConcreteNativeAssetFr
...
+pub fn weigh_multi_assets_generic(
+	filter: &MultiAssetFilter,
+	weight: Weight,
+	max_assets: u32,
+) -> XCMWeight {
+	let weight = match filter {
+		MultiAssetFilter::Definite(assets) => weight.saturating_mul(assets.len() as u64),
+		MultiAssetFilter::Wild(_) => weight.saturating_mul(max_assets as u64),
+	};
+	weight.ref_time()
```
Comment (gilescope) [inline parachains/common/src/xcm_config.rs:134]: Could pull out the multiplier:
```suggestion
	let multiplier = match filter {
		MultiAssetFilter::Definite(assets) => assets.len() as u64,
		MultiAssetFilter::Wild(_) => max_assets as u64,
	};
	weight.saturating_mul(multiplier).ref_time()
```

**Review by gilescope — APPROVED** [review]
LGTM

File: parachains/runtimes/assets/statemine/src/weights/xcm/mod.rs [hunk 3]
```diff
@@ -31,22 +31,28 @@ trait WeighMultiAssets {
 	fn weigh_multi_assets(&self, weight: Weight) -> XCMWeight;
 }
 
-const MAX_ASSETS: u32 = 100;
+trait WeighMultiAssetsReserve {
+	fn weigh_multi_assets_reserve(&self, weight: Weight) -> XCMWeight;
+}
+
+const RESERVE_MAX_ASSETS: u32 = 100;
+const MAX_ASSETS: u32 = 1;
```
[the remark this answers is not in the document]
Comment (ruseinov) [reply parachains/runtimes/assets/statemine/src/weights/xcm/mod.rs:39]: Nope, because it also does deposits! I will add a comment instead

**Comment by ruseinov** [comment]
bot merge

**Comment by joepetrowski** [comment]
I really think the switch in https://github.com/paritytech/cumulus/pull/1639/commits/81ea3bec38a246e75bbbba2777526a7622576935 was not correct. For example, setting `max_assets` to 1 for `DepositAsset` will break the ability to pay fees with an array of sufficient assets.

File: parachains/runtimes/assets/statemine/src/weights/xcm/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (KiChjang) [reply parachains/runtimes/assets/statemine/src/weights/xcm/mod.rs:40]: This looks incorrect: both `DepositAsset` and `InitiateTeleport` uses this value, and if they ever use `Wild(All)` as their `MultiAssetFilter`, the weights given here will be incorrect, as it assumes that the code only interacts with a maximum of 1 asset.

**Pull request merged**